areader/reader.go

```diff
@@ -368,7 +359,13 @@ func (ar *Reader) readHeaderV3(tReader *tar.Reader,
 			return nil, errors.Wrap(err, "readHeaderV3")
 		}
 		parsePath = append(parsePath, hdr.Name)
-		nextParseToken, validPath := verifyParseOrder(parsePath)
+		nextParseToken, validPath, err := verifyParseOrder(parsePath)
+		if err != nil {
+			panic("The only error returned from verifyParseOrder should be errParseOrder")
+		}
+		if err == errParseOrder {
+			return nil, fmt.Errorf("Invalid structure: %s, wrong element: %s", parsePath, parsePath[len(parsePath)-1])
+		}
```

Comment:
Hmm, having a `panic` makes me a bit nervous, even though it is an invariant check. This code will run both on the client and in the server. Could we simply return the same error instead?

In any case, you probably meant to reverse the ifs, since you will get the panic for all errors if they stay in this order.

Reply:
:see_no_evil:  yeh, agreed.

Reply:
Btw, how do we then guarantee invariance? or should it simply be ignored?

Reply:
Well, we *do* guarantee invariance, it's just a different way of handling it. If we hit the check here then we return from the function, and outside the function the invariance is no longer relevant. We will thus fail to parse correctly, which is a grave bug in dire need of fixing, but at least your server won't crash and can (hopefully) still handle other duties.

Hard invariant asserts are in general a better fit for C, where such checks can expose that your memory is all messed up, which can be *really* bad. Truly messing up memory is much harder to achieve in Go, so a softer fallback can be used.